### ISSUE TITLE
Errno::E2BIG: Arg list too long

### ISSUE THREAD

**Issue by beckerzito** [ThrowTheSwitch/Ceedling#560]
Hey guys, 

Recently I've been working in a C project that has a long folder structure. With that, in my first compilation attempt, I got the error "Errno::E2BIG: Arg list too long" due to the big include argument list generated for the project.

My question is: Have you already faced it? How did you manage to solve that? 

I was thinking in some kind of tool to generate the include arguments and then pass to the GCC (compiler that I'm using). But it would require a new process on the building system to generate the includes file.

What do you think?

**Comment by beckerzito** [comment]
@mvandervoord help me out here! Is there any solution already done for that?

**Comment by Letme** [comment]
Maybe not related, but I wonder how linux does it... The link list is probably really long, so there might be some trick to avoid that.

Are you using windows? Is that error message from Ruby?

**Comment by beckerzito** [comment]
> Maybe not related, but I wonder how linux does it... The link list is probably really long, so there might be some trick to avoid that.
> 
> Are you using windows? Is that error message from Ruby?

Yes I'm using windows.
I'm no really sure, but I'm guessing that this is an issue from the command prompt. It happens right after the ruby calls the entire command with gcc.exe and its arguments, including the very long include directories.

**Comment by Letme** [comment]
Can you count number of arguments you have, and we can maybe get a limit?

**Comment by mvandervoord** [comment]
I think it's less about the number of arguments and more about total line length? (which is 8191 characters on Windows)

**Comment by beckerzito** [comment]
Yes, the point is that since the start of the command "gcc.exe" until the end of it with "-DTEST -DGNU_COMPILER", the complete command has almost 37.815 characters.

So, I see that this can be a pretty common issue to the tool. It's not too hard to find big projects on the industries and other applications, that could easily overflow the 8191 digits. 

Any thoughts on how to solve that? 
From my side, I'm trying to modify the ceedling in other to not pass the include directories through the command line, but by the environment variable 'CPATH' that could be a possibility to the GCC, but it doesn't seem like a global solution for other compilers. 

Let me know your comments!

**Comment by mvandervoord** [comment]
is that what most of the characters are in your setup? include directories?

**Comment by beckerzito** [comment]
Yes, the biggest part of the characters comes from the include directories.

**Comment by beckerzito** [comment]
Any suggestion on that guys?

**Comment by dpostorivo** [comment]
I've hit this before on linux when having globs in paths with a lot of directories. My gut is that we can give a smarter error for this, but it would take some edge testing. If you can trim down your include paths to be more specific it would take care of itself, but I agree ceedling should be a bit smarter and give a better error message.

**Comment by MichaelBMiner** [comment]
I have the same issue trying to use ceedling for an embedded linux project. All of the includes generate a line that is 200K characters long. 

Have you found a solution for this? Is there a way to keep my includes and run ceedling?

**Comment by serjche** [comment]
Hi, I'm in a similar situation, which I didn't account for when establishing ceedling for a big project.
Additionally to what has been said in this issue, I observed that the whole list of all the values defined in the :paths: variable in the project.yml file, are always included in all the operations with gcc, regardless if they pertain to the test or not.

For example, let's say that we have two test files in different folders
- driver1/unit_test/test_driver1.c
- driver2/unit_test/test_driver2.c

And a file structure such as:

    .
    ├── driver1
    │   ├── inc
    │   ├── src
    │   └── unit_test
    ├── driver2
    │   ├── inc
    │   ├── src
    │   └── unit_test

Where the inc and src folders contain the .h and .c files, as you might expect.

Let's say that both drivers are completely independent of each other, they don't requiere headers from each other.

with a project.yml file such as 
```yaml
:paths:
  :tests:
    - driver1/unit_test
    - driver2/unit_test
  :source:
    - driver1/src
    - driver2/src
  :include:
    - driver1/inc
    - driver2/inc
```

When running `ceedling verbosity[4] test:all` I found that for each test ran, the preprocess/compilation/linker steps all have the whole list of includes (omitting other gcc options):
`gcc.exe ... -I"driver1/inc" -I"driver1/src" -I"driver1/unit_test" -I"driver2/inc" -I"driver2/src" -I"driver2/unit_test" ... test_driver1.c`

Even though test_driver1 does not need all the includes from driver2, but it does anyway.

This of course, does not affect in small projects, but on bigger ones, as some of you folks have mentioned, provoking the **Errno::E2BIG** error.

I was trying the CPATH approach by @beckerzito, but had no success so far.

Have any of you reached a solution or a workaround to this predicament?

This issue seems to be more related to the way Ceedling organizes all the include paths in every operation it makes, so I think it is a more fundamental problem that we might think at first.

If there is no known workaround, do you guys have any proposal or idea on how we could implement a way in which only certain paths are included exclusively for certain tests.

I would really appreciate your help. I'll keep trying to find a work around in the meantime.

Thank you!

**Comment by SilverBulletEngineering** [comment]
I suggest you make two separate projects (yml files).  It's easy to switch between projects.

**Comment by bmcgrath-mp** [comment]
I am trying to use ceedling to build and run units tests for an esp32 project. The esp-idf/components/ directory has 2651 directories, 15354 files some of which are include files needed to build or at least be mocked.

I also get the "Errno::E2BIG: Arg list too long" if I try to include the esp-idf/components directory, and this is when building on the throwtheswitch docker image of ubuntu.

I tried to narrow it down a bit (see code snippet below) but still end up with "no such file or directory" type errors. Has anyone successfully solved this issue or can point me to an example of building unit tests for esp32 with ceedling? I don't need to use the esp-idf toolchain to build for target within ceedling, my plan is to keep unit tests on the host via ceedling and target builds via esp-idf. I don't want to use espressif's half baked linux example either. Thanks!

if it helps this is what my project.yml is turning into

```:include:
    - +:/esp-idf/*/include/**
    # the below shouldn't be necessary but "ceedling files:include" output shows they are omitted if not included explicitly...
    - +:/esp-idf/esp_system/include/*
    - +:/esp-idf/esp_common/include/*
    - +:/esp-idf/soc/**

**Comment by jnz86** [comment]
I'm disappointed this still isn't fixed. It seems like it's probably pretty common for anyone using larger projects in windows with it's long paths. 

Trying the work around. But the silence on the issues here is really a bummer.

**Comment by Letme** [comment]
This is not a Ceedling problem - this is Windows problem. 

You have some workarounds on top, but I think best advice is: switch to Linux. Much more developer friendly.

**Comment by serjche** [comment]
@Letme, really? that's not very helpful on your part.

The error itself is from GCC itself, not windows, although the windows build of GCC might be more susceptible to it, some folks on this very thread said that this has happened also on linux, and on the following issue it is reported for linux as well: https://github.com/ThrowTheSwitch/Ceedling/issues/480

The main reason this happens is the way ceedling takes every single path in the .yaml configuration and appends it to every single compilation step for every single file, which may be unnecessary for most of the test files.

It is a fundamental issue on how ceedling works, but a proper solution might be quite cumbersome to produce, to only include the necessary -I paths on each compilation step (I'm not really an expert on the project, take that with a grain of salt).

@jnz86 were you able to solve the issue with the workaround?

**Comment by jnz86** [comment]
Yes and No.

I used the exact workaround and still had the error, but before double checking anything. I realized that Ceedling was compiling very slowly, so I removed the all inclusive source line from my project. I much more narrowly include files now, its faster and I'm no longer at the limit.

However, I agree with you that this isn't really a solution and if I keep using this I'll hit the limit again. I've kicked the can down the road a little bit. 

The workaround may be flawless and I messed up somewhere else. I have at least two installs of Ruby and both have Ceedling.
